source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/PowerShellBootstrapper.cs

```diff
@@ -65,37 +71,63 @@ public override string PathToPowerShellExecutable()
+                    .Select<string, (string path, string versionId, int? majorVersion, string remaining)>(d =>
                     {
                         var directoryInfo = new DirectoryInfo(d);
                         var directoryName = directoryInfo.Name;
-			
+
                         // Directories are typically versions like "6" or they might also have a prerelease component like "7-preview"
-                        var splitString = directoryName.Split(new[] { '-' }, 2);
+                        var splitString = directoryName.Split(new[] {'-'}, 2);
                         var majorVersionPart = splitString[0];
-                        var preRelease = splitString.Length < 2 ? string.Empty : splitString[1]; // typically a prerelease tag, like "preview"
-			
+                        var preRelease =
+                            splitString.Length < 2
+                                ? string.Empty
+                                : splitString[1]; // typically a prerelease tag, like "preview"
+
                         if (int.TryParse(majorVersionPart, out var majorVersion))
-                            return (d, majorVersion, preRelease);
-                        return (d, null, preRelease);
-                    })
+                            return (d, directoryName, majorVersion, preRelease);
+                        return (d, directoryName, null, preRelease);
+                    }).ToList();
+                var latestPowerShellVersionDirectory = availablePowerShellVersions
+                    .Where(p => string.IsNullOrEmpty(customVersion) || p.versionId == customVersion)
                     .OrderByDescending(p => p.majorVersion)
                     .ThenBy(p => p.remaining)
-                    .First()
-                    .path;
+                    .Select(p => p.path)
+                    .FirstOrDefault();
+                
+                if (customVersionIsDefined && latestPowerShellVersionDirectory == null)
+                {
+                    throw new PowerShellVersionNotFoundException(customVersion, availablePowerShellVersions.Select(v => v.versionId));
+                }
+
                 var pathToPwsh = Path.Combine(latestPowerShellVersionDirectory, EnvPowerShellPath);
 
-                powerShellPath = File.Exists(pathToPwsh) ? pathToPwsh : EnvPowerShellPath;
+                return File.Exists(pathToPwsh) ? pathToPwsh : EnvPowerShellPath;
+            }
+            catch (PowerShellVersionNotFoundException)
+            {
+                throw;
             }
             catch (Exception)
             {
-                powerShellPath = EnvPowerShellPath;
+                return EnvPowerShellPath;
             }
+        }
 
-            return powerShellPath;
+        protected override IEnumerable<string> ContributeCommandArguments(CalamariVariableDictionary variables)
+        {
+            yield break;
```

Comment:
What does this do?

Reply:
Remind me at our next catch up and I'll run you through it